### ISSUE TITLE
Automatically add exported variable as global in importing file

### ISSUE THREAD

**Issue by TimvdLippe** [eslint/eslint#6325]
**What version of ESLint are you using?**
v2.11.1

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**The problem you want to solve.**
Whenever I set a variable to be `/**exported VARNAME*/` I need to redeclare this `/**global VARNAME*/` in every file that uses this file. E.g. I put an exported variable bar (`/**exported bar*/`) in `foo.js`. Whenever I include `<script src="foo.js"></script>` I have to redeclare `/**global bar*/` in this file.

Whenever I change the name of the variable or use bar anywhere else, I have to update/add this global annotation.

**Your take on the correct solution to problem.**
It would greatly improve the readability and maintenance if this would be automatically deduced from the existing source code annotations.

**Comment by platinumazure** [comment]
Hmm... there are a few constraints imposed by the design and architecture of ESLint that might make this less feasible:

1. ESLint operates on one file at a time and our core rules will not perform filesystem operations to pull in imported files.
1. ESLint operates on files in an arbitrary order, so it would be tough to store exported variables in a dictionary and hope that the reading file comes later in the linting process.
1. Doing a second pass on all the files would probably be prohibitively expensive.

Given these constraints, can you see a way of implementing your proposal in ESLint? Please note that even if we can't have this in core, you can always write a custom rule or plugin which does what you need (probably using a filesystem-based approach, since file traversal is controlled by ESLint core).

**Comment by TimvdLippe** [comment]
Sadly I have no experience with ESLint core, I am merely an end-user. My programmer intuition thinks that a post-process step removing all "incorrect" rules would be the most performant. I understand the architectural limitations, maybe this feature is possible in a later version of ESLint where such features can be implemented?

**Comment by ilyavolodin** [comment]
This request goes again a few of our architecture principles. First of all, as @platinumazure mentioned, we are working on a single file at a time, so we can't share information from one file to the other. Second, we rely on globs to tell us which files to lint, and we don't control order of linting at all. Because of that, we have absolutely not guarantee that the file that would export variables will be linted before file that imports them. Also, all of our rules are completely independent and don't know anything about each other (and because of that can't sure any information), so even though in this case the rule would be sharing information with itself, it's still going to be two different instances of the same rule.

In short, I don't think we would be able to justify major architectural changes that would be required to enable this feature. This would require us to more or less re-write the whole tool. Sorry.

Closing, as I don't think this is something that we would ever be able to implement correctly. Feel free to reopen if you disagree.
